flink/flink-1.13/src/test/java/io/tidb/bigdata/flink/tidb/FlinkTest.java

```diff
@@ -430,4 +386,95 @@ public void testSnapshotRead() throws Exception {
+        String.format("CREATE TABLE `%s` (`c1` int unique key)", tableName),
+        String.format("SPLIT TABLE `%s` BETWEEN (0) AND (2000) REGIONS %s", tableName, 2),
+        String.format("INSERT INTO `%s` VALUES %s", tableName, values));
+    EnvironmentSettings settings = EnvironmentSettings.newInstance().useBlinkPlanner()
+        .inStreamingMode().build();
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.enableCheckpointing(100L);
+    StreamTableEnvironment tableEnvironment = StreamTableEnvironment.create(env, settings);
+    env.setParallelism(1);
+    properties.put("type", "tidb");
+    String createCatalogSql = format("CREATE CATALOG `tidb` WITH ( %s )",
+        TableUtils.toSqlProperties(properties));
+    tableEnvironment.executeSql(createCatalogSql);
+    String queryTableSql = format("SELECT * FROM `%s`.`%s`.`%s`", "tidb", DATABASE_NAME,
+        tableName);
+    Table table = tableEnvironment.sqlQuery(queryTableSql);
+    SingleOutputStreamOperator<RowData> dataStream = tableEnvironment.toAppendStream(
+        table, RowData.class).map(row -> {
+      if (RandomUtils.randomInt(1000) == 0) {
+        throw new IllegalStateException("Random test exception");
+      }
+      Thread.sleep(10L);
+      return row;
+    });
+    tableEnvironment.createTemporaryView("test_view", dataStream);
+    CloseableIterator<Row> iterator = tableEnvironment.executeSql("SELECT * FROM `test_view`")
+        .collect();
+    Set<Integer> results = new HashSet<>();
+    while (iterator.hasNext()) {
+      Row row = iterator.next();
+      results.add(((RowData) row.getFieldAs(0)).getInt(0));
+    }
+    Assert.assertEquals(2000, results.size());
+    Assert.assertEquals(IntStream.range(0, 2000).boxed().collect(Collectors.toSet()), results);
+  }
+
+  @Test
+  public void testReadExactlyOnce() throws Exception {
+    Map<String, String> properties = ConfigUtils.defaultProperties();
+    ClientSession clientSession = ClientSession.create(new ClientConfig(properties));
+    String tableName = RandomUtils.randomString();
+    String values = IntStream.range(0, 2000).mapToObj(i -> "(" + i + ")")
+        .collect(Collectors.joining(","));
+    clientSession.sqlUpdate(
+        String.format("CREATE TABLE `%s` (`c1` int unique key)", tableName),
+        String.format("SPLIT TABLE `%s` BETWEEN (0) AND (2000) REGIONS %s", tableName, 2),
+        String.format("INSERT INTO `%s` VALUES %s", tableName, values));
+    EnvironmentSettings settings = EnvironmentSettings.newInstance().useBlinkPlanner()
+        .inStreamingMode().build();
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.enableCheckpointing(100L);
+    StreamTableEnvironment tableEnvironment = StreamTableEnvironment.create(env, settings);
+    env.setParallelism(1);
+    properties.put("type", "tidb");
+    properties.put(TiDBOptions.SOURCE_FAILOVER.key(), "offset");
```

Comment:
Do we need test the default value

Reply:
See testReadAtLeastOnce(), the method above this